File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int len = nums.length;
        k = k % len;
        int[] temp = new int[len];

        for (int i = 0; i < len; i++) {
            temp[(i + k) % len] = nums[i]; // Shift elements by k positions
        }

        System.arraycopy(temp, 0, nums, 0, len); // Copy temp back to nums
    }
}

File: 0189-rotate-array/0189-rotate-array.py
from typing import List

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        k = k % len(nums)  # Handle cases where k > len(nums)
        
        def reverse(start, end):
            while start < end:
                nums[start], nums[end] = nums[end], nums[start]
                start += 1
                end -= 1
        
        # Step 1: Reverse the entire array
        reverse(0, len(nums) - 1)
        
        # Step 2: Reverse the first k elements
        reverse(0, k - 1)
        
        # Step 3: Reverse the remaining elements
        reverse(k, len(nums) - 1)
